File: MotionEditor/src/de/naoth/me/core/Scanner.java
package de.naoth.me.core;

import java.io.*;


/**
 *
 * @author claas
 */
public class Scanner
{
    private Character LineComment = '#';
    private Character nextChar = ' ';
    //private int bufferLength = 0;
    private int lineNumber = 1;
    public String buffer = "";
    private Reader inputStream = null;

    //Constructor
    public Scanner(Reader input)
    {
        inputStream = input;
    }
    //end Constructor

    public void setLineCommentChar(Character LineCommentChar)
    {
      this.LineComment = LineCommentChar;
    }
    
    public boolean endOfFile()
    {
        return eof();
    }
    //end endOfFile

    public void getNextToken()
    {
      skipSpace();
      //bufferLength = 0;
      buffer = "";

      if(eof())
      {
        buffer += nextChar;
        //buffer[bufferLength++] = nextChar;
        //buffer[bufferLength] = '\0';
        return;
      }//end if

      if(!isTokenChar(nextChar))
      {
        buffer += nextChar;
        //buffer[bufferLength++] = nextChar;
        //buffer[bufferLength] = '\0';
        getNextChar();
        return;
      }//end if

      while(isTokenChar(nextChar))
      {
        buffer += nextChar;
        //buffer[bufferLength++] = nextChar;
        //buffer[bufferLength] = '\0';
        getNextChar();
      }//end while
    }
    //end getNextToken


    /*
     * Return true if c is a token character,
     * i.e., 0-1, a-z, A-Z, or one of the character '.','-' or '_'
     */
    public boolean isTokenChar(char c)
    {
      return Character.isLetterOrDigit(c) || c == '.' || c == '-' || c == '_';
    }
    //end isTolenChar


    public boolean isToken(String token)
    {
//      System.out.println("a:" + (buffer.compareTo(token) == 0) + ":e");
      return buffer.compareTo(token) == 0;
    }
    //end isToken

    private boolean eof()
    {
      return nextChar == -1; //EOF
    }
    //end eof

    private void skipSpace()
    {
      while(Character.isWhitespace(nextChar) || (nextChar == LineComment && LineComment != null) )
      {
        if(nextChar == '\n')
          lineNumber++;

        if(nextChar == LineComment && LineComment != null)
          skipLine();
        else
          getNextChar();
      }
    }
    //end skipSpace


    private void skipLine()
    {
      while(!eof() && nextChar != '\n')
      {
        getNextChar();
      }//end while

      getNextChar();
      lineNumber++;
    }
    //end skipLine


    public void getNextChar() 
    {
      try
      {
        if(!eof())
          nextChar = (char) inputStream.read();
      }
      catch (IOException e)
      {
        nextChar = (char) -1;
      }
      //else
        //nextChar = -1; //EOF
    }
    //end getNextChar


    public int getLineNumber()
    {
      return lineNumber;
    }
    //end getLineNumber
}

File: NaoTHSoccer/Source/Motion/Engine/KeyFrameMotion/Tools/Scanner.cpp
/**
* @file Scanner.h
*
* @author <a href="mailto:mellmann@informatik.hu-berlin.de">Heinrich Mellmann</a>
* Implementation of class Scanner
*/

#include "Scanner.h"

bool Scanner::endOfFile()
{
  return nextChar == EOF;
}//end eof


void Scanner::getNextToken()
{
  skipSpace();
  bufferLength = 0;

  if(nextChar == EOF) 
  {
    buffer[bufferLength++] = nextChar;
    buffer[bufferLength] = '\0';
    return;
  }//end if

  if(!isTokenChar(nextChar))
  {
    buffer[bufferLength++] = nextChar;
    buffer[bufferLength] = '\0';
    getNextChar();
    return;
  }//end if

  while(isTokenChar(nextChar))
  {
    buffer[bufferLength++] = nextChar;
    buffer[bufferLength] = '\0';
    getNextChar();
  }//end while
}//end getNextToken

/*
 * Return true if c is a token character, 
 * i.e., 0-1, a-z, A-Z, or one of the character '.','-' or '_'
 */
bool Scanner::isTokenChar(char c)
{
  return isalnum(c) || c == '.' || c == '-' || c == '_';
}//end isTolenChar


bool Scanner::isToken(std::string token)
{
  return token == buffer;
}//end isToken

bool Scanner::eof()
{
  return nextChar == EOF;
}//end eof

void Scanner::skipSpace()
{
  while(isspace(nextChar) || nextChar == '#')
  {
    if(nextChar == '\n')
      lineNumber++;

    if(nextChar == '#') 
      skipLine();
    else
      getNextChar();
  }
}//end skipSpace


void Scanner::skipLine()
{
  while(nextChar != EOF && nextChar != '\n')
  {
    getNextChar();    
  }//end while
  
  getNextChar(); 
  lineNumber++;
}//end skipLine


void Scanner::getNextChar()
{
  if(inputStream.good())
    nextChar = (char) inputStream.get();
  else
    nextChar = EOF;
}//end getNextChar


int Scanner::getLineNumber()
{
  return lineNumber;
}//end getLineNumber


